### PR TITLE
Determine a branch target of DBNZ correctly (version 1)

### PR DESCRIPTION
**Pull request by kolerov**
DBNZ instruction is in BRANCH class as other branch instructons like B, Bcc, etc. However, DBNZ is the only one branch instruction
that stores a branch offset not in the first operand but in the second one.

However, current logic of arc_insn_get_branch_target assumes that a branch offset is always stored in the first operand for BRANCH class. Thus, it's necessary to add extra logic for determining DBNZ and getting a correct offset.

This PR resolves this issue: https://github.com/foss-for-synopsys-dwc-arc-processors/toolchain/issues/546

### CODE REVIEW

File: gdb/testsuite/gdb.arch/arc-dbnz.S
Comment (shahab-vahedi) [inline gdb/testsuite/gdb.arch/arc-dbnz.S:1]: This test is beautiful, but please also add `.d` to at least one of them `dbnz`s.

File: gdb/testsuite/gdb.arch/arc-dbnz.exp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+# Test dbnz instruction. It decrements its source register operand, and if
+# the result is non-zero it branches to the location defined by a signed
+# half-word displacement operand.
+#
+# It's necessary to verify these cases:
+#
+#     1. Check that dbnz does not branch if its source register is 0 
+#        after decrementing.
+#     2. Check that dbnz branches forward correctly if its source register
```
Comment (shahab-vahedi) [inline gdb/testsuite/gdb.arch/arc-dbnz.exp:26]: As you've shown in your test, `dbnz` does not necessarily branch _forward_ or _backward_. Respectively, it either _falls through_ or _jumps to the branch target_. Please adapt points "2 and 3" accordingly AND do the same for the commit message.

File: gdb/arc-tdep.c [hunk 2]
```diff
@@ -465,6 +465,16 @@ arc_insn_get_branch_target (const struct arc_instruction &insn)
 	 instruction, hence last two bits should be truncated.  */
       return pcrel_addr + align_down (insn.address, 4);
     }
+  /* DBNZ is the only branch instruction that keeps a branch address in
+     in the second operand. It must be intercepted and treated differently. */
```
Comment (shahab-vahedi) [inline gdb/arc-tdep.c:469]: typo: duplication of the word `in`:

`in the second ...` -> `the second ...`

File: gdb/arc64-tdep.c [hunk 3]
```diff
@@ -419,6 +419,16 @@ arc_insn_get_branch_target (const struct arc_instruction &insn)
 	 instruction, hence last two bits should be truncated.  */
       return pcrel_addr + align_down (insn.address, 4);
     }
+  /* DBNZ is the only branch instruction that keeps a branch address in
+     in the second operand. It must be intercepted and treated differently. */
```
Comment (shahab-vahedi) [inline gdb/arc64-tdep.c:423]: typo: duplication of the word `in`:

`in the second ...` -> `the second ...`

**Review by shahab-vahedi — APPROVED** [review]
With only a few nit-picks fixed.

**Comment by kolerov** [comment]
I created 2 different PR from this:

1. https://github.com/foss-for-synopsys-dwc-arc-processors/binutils-gdb/pull/77
2. https://github.com/foss-for-synopsys-dwc-arc-processors/binutils-gdb/pull/78

**Pull request closed without merging**